### PR TITLE
add `maxAge` option for `alfy.cache` - fixes #3

### LINKED ISSUE #3: Add `maxAge` option for `alfy.cache`
// @SamVerschueren

### PR DESCRIPTION
**Pull request by SamVerschueren**
- Extended the `Conf` class and override the `set`, `get` and `has` methods
- Added tests for the cache
- Added documentation regarding the `maxAge` option

As described in #3, not sure if we should do it with the `CacheConf` object here, add that option to `Conf` or create an external `CacheConf` module to handle this. Will to do either one of those :).

Feel free to suggest improvements of the code or docs.

### CODE REVIEW

**Comment by SamVerschueren** [comment]
The downside of this approach is that it isn't possible to implement a fallback mechanism like I do in [alfred-ng2](https://github.com/SamVerschueren/alfred-ng2/blob/master/index.js#L32-L39). If the cache is expired, it tried to load the remote data, if that fails it will still load the "old" cache. It would be nice if it was possible to still support that.

-------------------

It actually is possible if you use `alfy.cache.store` directly, but I don't want everyone to skip the methods. Maybe we should not remove the item if the data is expired and add an extra option to get in order to force retrieval.

```js
alfy.cache.set('foo', 'bar', {maxAge: 5000});

await delay(5000);

alfy.cache.get('foo', true);
//=> 'bar'
```

**Comment by sindresorhus** [comment]
> The downside of this approach is that it isn't possible to implement a fallback mechanism like I do in alfred-ng2. If the cache is expired, it tries to load the remote data, if that fails it will still load the "old" cache. It would be nice if it was possible to still support that.

Hmm, maybe we should bundle `got` here too, and give it ability to cache requests? That would solve your use-case.

**Comment by SamVerschueren** [comment]
Not sure if we should bundle `got` into `alphy`. How would the interface look like then?

```js
alfy.got.get('https://myapi.com/foo/bar.json', {maxAge: 5000}).then(res => {
    console.log(res.body);
    // {foo: 'bar'}
});
```

**Comment by sindresorhus** [comment]
@SamVerschueren Fetching from network is a very common operation in Alfred workflows. Would be nice to simplify it. Dependencies are super cheap anyways as they're bundled with the workflow, not `npm install`'d.

I'm thinking:

```js
alfy.fetch('https://myapi.com/foo/bar.json', {maxAge: 5000}).then(res => {
	console.log(res.body);
	// {foo: 'bar'}
});
```

And maybe even make `{json: true}` the default.

-

To be clear, this is in addition what's currently in this PR, not instead of. Should probably be a separate PR.

**Comment by SamVerschueren** [comment]
So we would only expose the `get` method right? Don't think that `post` makes sense in an Alfred app. Except if the API endpoint expects a `post` request with `alfy.input` as post data. Nah, let's just expose the `get` as `fetch`. If they have to use post, they can always install `got` and use `got` directly instead of via `alfy`.

> Dependencies are super cheap anyways as they're bundled with the workflow

👍

**Comment by sindresorhus** [comment]
> So we would only expose the get method right?

Yup.

> Except if the API endpoint expects a post request with alfy.input as post data.

Then they can just pass the `method` option. We just pass the options object to `got`.

> Is it ok to remove the cached data then when it is expired? Or should we just ignore it without removing it?

For `alfy.cache`, remove it. For `alfy.fetch`, remove it after successfully getting new data.

**Comment by SamVerschueren** [comment]
> Then they can just pass the method option. We just pass the options object to got.

👍 

> For alfy.cache, remove it. For alfy.fetch, remove it after successfully getting new data.

👍

File: readme.md [hunk 1]
```diff
@@ -154,6 +154,22 @@ alfy.cache.get('unicorn');
...
+##### maxAge
+
+The `set` method of this instance accepts an optional third argument where you can provide a `maxAge` option. `maxAge` is
+the number of milliseconds the value is valid in the cache.
+
+Example:
+
+```js
+alfy.cache.set('foo', 'bar', {maxAge: 5000});
+
```
Comment (sindresorhus) [inline readme.md]: Do a `alfy.cache.get('foo');` and show the value before the delay too. For clarity.

File: readme.md [hunk 2]
```diff
@@ -154,6 +154,22 @@ alfy.cache.get('unicorn');
...
+await delay(5000);
+
+alfy.cache.get('foo');
+// undefined
```
Comment (sindresorhus) [inline readme.md]: `//=>`

File: readme.md [hunk 3]
```diff
@@ -154,6 +154,22 @@ alfy.cache.get('unicorn');
...
+
+The `set` method of this instance accepts an optional third argument where you can provide a `maxAge` option. `maxAge` is
+the number of milliseconds the value is valid in the cache.
+
+Example:
+
+```js
+alfy.cache.set('foo', 'bar', {maxAge: 5000});
+
+await delay(5000);
```
Comment (sindresorhus) [inline readme.md]: Users wouldn't know what `delay` is. I would do `await require('delay')(5000);` to be explicit where it came from.

File: readme.md [hunk 4]
```diff
@@ -154,6 +154,28 @@ alfy.cache.get('unicorn');
...
 
+##### maxAge
+
+The `set` method of this instance accepts an optional third argument where you can provide a `maxAge` option. `maxAge` is
+the number of milliseconds the value is valid in the cache.
+
+Example:
+
+```js
+const delay = require('delay');
```
Comment (SamVerschueren) [inline readme.md]: I prefer a require at the top instead of `await require('delay')(5000)`. But if you want it inline, I'll fix it :).

**Comment by sindresorhus** [comment]
Looks superb, as always :)

**Comment by SamVerschueren** [comment]
Awesome. Super nice to see this land :)! If #6 is implemented, I nearly don't have to do anything anymore in my `alfred-ng2` plugin 😆 .

**Pull request merged**